src/allmydata/interfaces.py

```diff
@@ -2936,3 +2936,68 @@ class IConnectionStatus(Interface):
         (pending, connected, refused, or other errors).
         """)
 
+
+class ITahoeUsagePolicy(Interface):
+    """
+    Callbacks in this interface are called by core Tahoe-LAFS APIs,
+    thus controlling whether particular API calls are allowed or not.
+
+    There should be enough information and hook points to support at
+    least these use-cases:
+
+      - anonymous, everything allowed (current default)
+      - "storage club" / "friend-net" (possibly identity based?)
+      - cryptocurrencies (ideally, paying for each API call?)
+      - anonymous tokens (payment for service, but without identities)
+    """
+
+    @property
+    def name():
+        """
+        How this policy shall be referred to (e.g. in Storage Server
+        announcements, to users, etc). Ideally limit yourself to
+        lowercase letters, numbers (but not a leading number) and
+        underscores.
+        """
```

Comment:
I think attributes on zope interfaces should be defined using `zope.interface.Attribute`.

Also, maybe this is "human_readable_name".  I think we also need some kind of globally unique identifier for programmatic references.  I've been thinking about using URLs in the way XML DTDs use them.  A couple examples I've played with so far are "https://tahoe-lafs.org/authz/anonymous/v1" (which would be for the current behavior) and "https://privatestorage.io/authz/SAT/v1" (for the plugin privatestorage plans to implement).

And if we don't want this to be arbitrary unicode then maybe we should be more precise about the constraints.  What are the motivations for restricting it?

Reply:
Arbitrary "worries" about using these in config files, Python code, etc. Like, with a URL is the .ini syntax you put in below even legal? A "human_readable_name" with restrictions can still be unique (there aren't going to be TONS of these things, and even just note to check a wiki with "here are current names, please don't re-use one" should work..?)

In XML, the URLs are actually functional, right? Like, I can fetch them and get some kind of schema or so. Maybe you're imagining the same for the above? (But, without that feature, I don't really see any advantage of them being real URLs vs. "a unique name")

Reply:
> In XML, the URLs are actually functional, right? Like, I can fetch them and get some kind of schema or so.

You basically never should.  I wasn't imagining these URLs would need to have a real resource associated with them.  It's more like it's an established string format that clearly demarcates ownership and provides for some hierarchy.  Nobody should be using any tahoe-lafs.org urls except the tahoe project itself, etc.  I'm not tied to this idea but I think something more than just "freeform human-facing text" is a good idea and isn't really any harder than leaving it totally open.

I think there is value in separating the identifier from the human-facing part.  You may want to tweak the human-facing part to improve discoverability, user experience, etc, without invalidating existing configurations.

> Like, with a URL is the .ini syntax you put in below even legal?

Good question!  Python's ConfigParser seems okay with it, at least.  I have no idea if there's some kind of official ini spec.  I imagine that if Tahoe-LAFS ever stops using Python's ConfigParser module, though, it will be to switch to a completely different syntax.